File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun goodNodes(root: TreeNode?): Int {
        val num = root?.`val` ?: 0

        return dfs(root, num)
    }

    private fun dfs(root: TreeNode?, maxSoFar : Int) : Int {
        if(root == null)
            return 0

        val isGreater = if(root.`val` >= maxSoFar) { 1 }
        else { 0 }

        val newMax = max(maxSoFar, root.`val`)

        val left = dfs(root.left, newMax)
        val right = dfs(root.right, newMax)

        return isGreater + left + right
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-0.cpp
class Solution {
    std::vector<int> memo;
    int size; 

    int dp(const int& num) {
        if(num == 1)
            return 1; 
        if(num == 2)
            return 2; 

        if(memo[num] != -1)
            return memo[num]; 

        memo[num] = dp(num - 1) + dp(num - 2); 

        return memo[num]; 
    }

public:
    int climbStairs(int n) {
        size = n; 
        memo = vector(n + 1, -1); 

        return dp(n);
    }
};
